src/Bicep.Core/Parsing/Parser.cs

```diff
@@ -1,370 +1,184 @@
-                : GetSuppressionFlag(ifCondition ?? assignment);
-
-            var body = this.WithRecovery(this.Object, suppressionFlag, TokenType.NewLine);
-
-            return new ModuleDeclarationSyntax(keyword, name, path, assignment, ifCondition, body);
-        }
-
-        private Token? NewLineOrEof()
-        {
-            if (reader.Peek().Type == TokenType.EndOfFile)
-            {
-                // don't actually consume the token
-                return null;
-            }
-
-            return NewLine();
-        }
-
-        private Token NewLine()
-        {
-            return Expect(TokenType.NewLine, b => b.ExpectedNewLine());
-        }
-
-        public SyntaxBase Expression(bool allowComplexLiterals)
-        {
-            var candidate = this.BinaryExpression(allowComplexLiterals);
-
-            if (this.Check(TokenType.Question))
-            {
-                var question = this.reader.Read();
-                var trueExpression = this.Expression(allowComplexLiterals);
-                var colon = this.Expect(TokenType.Colon, b => b.ExpectedCharacter(":"));
-                var falseExpression = this.Expression(allowComplexLiterals);
-
-                return new TernaryOperationSyntax(candidate, question, trueExpression, colon, falseExpression);
-            }
-
-            return candidate;
-        }
-
-        private SyntaxBase BinaryExpression(bool allowComplexLiterals, int precedence = 0)
+            static RecoveryFlags ConvertFlags(bool suppress) => suppress ? RecoveryFlags.SuppressDiagnostics : RecoveryFlags.None;
+
+            /*
+             * When we have an incomplete declarations like "param\n",
+             * the keyword is parsed but all other properties are set to 0-length SkippedTriviaSyntax or MalformedIdentifierSyntax
+             * to prevent stacking multiple parse errors on a 0-length span (which is technically correct but also confusing)
+             * we will only leave the first parse error
+             */
+            switch (precedingNode)
+            {
+                case IdentifierSyntax identifier when identifier.IsValid == false:
+                    return ConvertFlags(identifier.Span.Length == 0);
+
+                case SkippedTriviaSyntax skipped:
+                    return ConvertFlags(skipped.Span.Length == 0);
+
+                default:
+                    return RecoveryFlags.None;
+            }
+        }
+
+        private SyntaxBase TargetScope(IEnumerable<SyntaxBase> leadingNodes)
+        {
+            var keyword = ExpectKeyword(LanguageConstants.TargetScopeKeyword);
+            var assignment = this.WithRecovery(this.Assignment, RecoveryFlags.None, TokenType.NewLine);
+            var value = this.WithRecovery(() => this.Expression(allowComplexLiterals: true), RecoveryFlags.None, TokenType.NewLine);
+
+            return new TargetScopeSyntax(leadingNodes, keyword, assignment, value);
+        }
+
+        private SyntaxBase Decorator()

```

Comment:
Please update grammar.md with the changes

Reply:
Sure will do.

Reply:
Done.